### PR TITLE
Feature/qed couplings

### PR DESCRIPTION
**Pull request by niclaurenti**

### CODE REVIEW

File: src/eko/couplings.py
Comment (felixhekhorn) [reply src/eko/couplings.py:64]: I think between the two sectors there will always be a symmetry, with just the references swapped and beta adjusted

File: src/eko/couplings.py
Comment (AleCandido) [inline src/eko/couplings.py:480]: At the moment the cache is just empty, so I'm pretty sure you want to fill it.

In case the content will be a structured object, pay attention to the difference between a deep and shallow copy. In any case, sometimes even a shallow copy is just fine, but you have to be aware and do it on purpose. If it's not structured, never mind.

File: src/eko/beta.py
Comment (felixhekhorn) [inline src/eko/beta.py:109]: Can we make the names more consistent? e.g. `beta_as2_aem1` (or all underscore?)

Comment (AleCandido) [reply src/eko/beta.py:109]: There is a reason for this name choice: this is the beta function for as, thus `beta_as`, while if it were for a it would have been called `beta_a`. After the second underscore, it follows the order as a series of numbers, each prefixed with the coupling, but the first one, for which it is implicit (explicitly it would be `beta_as_as2aem1`).
You can decide to put the order on the other side of the underscore, but it is not more consistent.

Think about the two cases with the same order:
- in your proposal, they would be `beta_as2_aem1` (the beta function for as) and `beta_aem1_as2` (the beta function for the qed coupling): this looks more like a mistake, why the two things are not the same?
- with current convention instead they are: `beta_as_2aem1` and `beta_a_1as2`, and if you truncate at the second underscore, you clearly read for what the beta function is, while the second piece is spelling out the order

It looks cleaner to me (even if from the point of view of the single one your own looks more natural), but I'm not especially keen in any. Both work, it's just a matter of taste at this point.

Comment (felixhekhorn) [reply src/eko/beta.py:109]: accepted - though the correct explicit name would be `beta_qcd_as2aem1`

Comment (AleCandido) [reply src/eko/beta.py:109]: I agree

File: src/eko/couplings.py
Comment (felixhekhorn) [inline src/eko/couplings.py:128]: is there anything shared between the (if-) branches? if not make them a seperate function each

**Comment by felixhekhorn** [comment]
I guess you can now attempt to merge `develop` here - an important change that happened meanwhile in `develop` that the `numba` usage has reverted to it's "original state", i.e. just-in-time compilation: for that all numba calls just need to be `nb.njit(cache=True)`

**Comment by niclaurenti** [comment]
> I guess you can now attempt to merge `develop` here - an important change that happened meanwhile in `develop` that the `numba` usage has reverted to it's "original state", i.e. just-in-time compilation: for that all numba calls just need to be `nb.njit(cache=True)`

So you mean that I should change all the arguments of njit for example like:
@nb.njit("f8(f8,u1,u1,f8)", cache=True) -> @nb.njit(cache=True)

File: src/eko/couplings.py
Comment (giacomomagni) [inline src/eko/couplings.py:446]: These two if are just the same as before, can you put them together?

File: src/eko/couplings.py
Comment (giacomomagni) [inline src/eko/couplings.py:404]: I agree, this should be true at this point

File: src/eko/couplings.py
Comment (giacomomagni) [inline src/eko/couplings.py:392]: Is the case `(0,0)` meaningful?

**Review by giacomomagni — COMMENTED** [review]
Hi @niclaurenti, 
I have added some minor comments. 

I noticed you haven't merged the most updated `develop` into this branch.
Just pull before merging here... (If you are using VS code there is this [extension](https://marketplace.visualstudio.com/items?itemName=mhutchie.git-graph) which is very useful)

Since this PR is quite big and introduced some changes in the notation, would you mind adding also some documentation and fix the names (`beta` ... ) also there? 
I think an explanation about the available pto expansions will be useful for our users.

File: src/eko/couplings.py
Comment (AleCandido) [reply src/eko/couplings.py:392]: Nope, `(0,0)` just means there is no evolution, so in principle there is no reason to run `eko`.

Nevertheless, we might decide to gracefully fallback on some trivial values (identity for the EKO, flat for running couplings, and related).

File: src/eko/couplings.py
Comment (niclaurenti) [reply src/eko/couplings.py]: No it's not. It's because previously the names where given factorizing a term as1 from beta_qcd and aem1 from beta_qed but then i reinserted it but i forgot to change the names inside the functions. I'm going to fix it

File: src/eko/couplings.py
Comment (niclaurenti) [reply src/eko/couplings.py:516]: In the code in the case (0,0) the reference values of a_s,a_em are returned (no evolution).

File: src/eko/couplings.py
Comment (felixhekhorn) [inline src/eko/couplings.py:322]: Check the docstring also in practice! I'm quite sure `a_em` will be interpreted as `{a_e}m` which is not what you want ...

File: src/eko/couplings.py
Comment (niclaurenti) [reply src/eko/couplings.py:85]: Maybe I dindn't understand what you mean. Here beta0, beta1, ecc are passed as parameters beacuse I wanted to use the same function for both QCD and QED since they have the same form but the beta functions are different

File: src/eko/couplings.py
Comment (felixhekhorn) [reply src/eko/couplings.py:40]: (of course here and below)

File: src/eko/couplings.py
Comment (felixhekhorn) [inline src/eko/couplings.py:47]: it's not the "beta function" but the first coefficient of the function (of course here and below)

File: src/eko/couplings.py
Comment (felixhekhorn) [inline src/eko/couplings.py:49]: A better name would be scaling log

File: src/eko/couplings.py
Comment (felixhekhorn) [inline src/eko/couplings.py:54]: please use a proper math name (like µ_r^2)

File: src/eko/couplings.py
Comment (felixhekhorn) [inline src/eko/couplings.py:65]: maybe write that this is the default expression found everywhere, e.g. in |APFEL| eq. ...

File: src/eko/couplings.py
Comment (felixhekhorn) [inline src/eko/couplings.py:95]: please mention that this is *not* the default expression, but instead is coming from |APFEL| eq. ...

File: src/eko/couplings.py
Comment (felixhekhorn) [inline src/eko/couplings.py:130]: - maybe we should add a bibtex entry á là "Luca Rottoli, private communication" and cite that one here
- also mention that is obtained via ordinary iterated solution of the RGE

File: src/eko/couplings.py
Comment (felixhekhorn) [reply src/eko/couplings.py]: actually, why? `order` is coming from the outside ...

File: src/eko/gamma.py
Comment (niclaurenti) [reply src/eko/gamma.py:13]: I was thinking that maybe it's more clear if we drop this convention and we just rename beta_as_2aem1 -> beta_qcd_as2aem1. @AleCandido @giacomomagni what do you think?

Comment (felixhekhorn) [reply src/eko/gamma.py:13]: fine with me - as said before this is actually the fully explicit name

File: src/eko/couplings.py
Comment (felixhekhorn) [reply src/eko/couplings.py:85]: I guess he meant: take as argument `ref, beta0, b1, lmu` and compute the actual `b1` in the calling function - this way you can *reuse* `b1` instead of *recompute*

File: src/eko/couplings.py
Comment (felixhekhorn) [inline src/eko/couplings.py:54]: Please don't follow me blindly :see_no_evil: use your own brain :brain:  I meant this more in a "global" sense - follow the notation which is already there: e.g. https://eko.readthedocs.io/en/latest/theory/pQCD.html you see the default notation is `\mu_R`; same with the "scaling log" - you know what it is and if ever you find a better name or explanation (like ratio between reference scale and target scale or log expansion parameter or ...) use it - us your physics intuition :rocket: ;-)

File: src/eko/gamma.py
Comment (AleCandido) [reply src/eko/gamma.py:13]: Essentially fine, but for `beta` actually `beta_as` it is even more explicit than `beta_qcd` (because it is the derivative of `as` in the first place), while `gamma_as` makes not so much sense, `gamma_qcd` is much more meaningful (since it's coming from the contribution of QCD interaction).

If you want to stick to only one of the two (because at the end the theory and the coupling are one-to-one), then prefer `qcd`, since `gamma_as` is way less meaningful.

Comment (giacomomagni) [reply src/eko/gamma.py:13]: I would go for `gamma_qcd` as well.

File: src/eko/couplings.py
Comment (niclaurenti) [reply src/eko/couplings.py:94]: Do you know where I can find a reference that I can cite?? In the Apfel paper and documentation i couldn't find anything

Comment (felixhekhorn) [reply src/eko/couplings.py]: mmm you're right - that makes me worry that it seems there is official statement ... when we first came across this (meaning implementing NNLO) we asked to the whole collaboration how to obtain this, but we only got alternative solutions ... I can see two options: or we trust APFEL and just cite it (without specifying a specific equation), or we fall back to the ordinary, iterated solution

File: src/eko/couplings.py
Comment (felixhekhorn) [inline src/eko/couplings.py:39]: Actually, this function should be called `exact_lo` as we can solve the RGE at LO exact and so we do always

File: src/eko/beta.py
Comment (felixhekhorn) [inline src/eko/beta.py:39]: Are you sure the Herzog paper discusses QED?

File: src/eko/couplings.py
Comment (AleCandido) [reply src/eko/couplings.py]: In lack of better references, cite APFEL

**Comment by felixhekhorn** [comment]
- @niclaurenti try to merge `develop` here (i.e. resolve the conflicts)
- please double check all LHA settings (including SV) are still passing
- then I guess we can merge ...
- just to remind everybody: this is a massive change both in code and interface and it will e.g. require a rewrite of all runcards (e.g. in pineko @andreab1997 )

**Comment by niclaurenti** [comment]
> What is still missing here? @niclaurenti @felixhekhorn
> 
> Otherwise I would merge, and most likely release v0.10.0 (such that the moment we'll be able to adapt yadism for this as soon as we want to, but it is not going to happen automatically with `poetry update`).

It is complete. I thought that you wanted to review it again before merging.

**Comment by felixhekhorn** [comment]
> Ok. Actually I put a function in compatibility.py that should interface the old runcards with the one that is used in this branch

Actually, this does not cover the shift of `PTO` -> `PTO+1` - maybe you should add this (and change all the benchmark cards again)

File: benchmarks/CT14_bench.py
Comment (felixhekhorn) [inline benchmarks/CT14_bench.py:66]: this has to be lifted inside `ekomark` otherwise the others (LHA, etc.) won't benefit from that - more specific it should move just before the `eko` invocation (since the DB is unchanged)

File: src/eko/evolution_operator/__init__.py
Comment (AleCandido) [inline src/eko/evolution_operator/__init__.py:289]: Maybe here and below I'd rename variables to `as0` and `as1`, such that it is clear at first inspection that we are still talking about the strong coupling.

File: src/eko/evolution_operator/grid.py
Comment (AleCandido) [inline src/eko/evolution_operator/grid.py:114]: Maybe make them explicitly integers as before (I don't remember if or which, but if `int` was there, there might have been a reason).
```suggestion
        config["order"] = tuple(int(o) for o in theory_card["order"])
```

File: src/eko/runner.py
Comment (AleCandido) [inline src/eko/runner.py:10]: Never use `eko` eplicitly inside `eko`. Use relative imports:
```suggestion
from . import compatibility
```

File: src/eko/evolution_operator/grid.py
Comment (AleCandido) [inline src/eko/evolution_operator/grid.py:34]: Even in this file, above, there is an absolute `import` from `eko`, please make it relative (as all the others)

**Review by AleCandido — CHANGES_REQUESTED** [review]
I reviewed everything but the actual `couplings.py` file and `compatibility.py`, I'll do as soon as possible.

File: src/eko/couplings.py
Comment (felixhekhorn) [inline src/eko/couplings.py:625]: I'd say let's provide some syntactical sugar: `def a_s(...): return a(...)[0]`

**Comment by AleCandido** [comment]
I didn't review line by line the whole `couplings.py`, but what I read was fine.

If you can fix the last two comments, I believe you can merge right after.

File: src/eko/compatibility.py
Comment (felixhekhorn) [reply src/eko/compatibility.py:30]: mh? but all old cards have a QED field - it was in APFEL in the first place

File: src/eko/compatibility.py
Comment (niclaurenti) [reply src/eko/compatibility.py]: Here I had put it since in the tests we don't always provide the operator card so I hat to define an empty card and pass it. But now it should be fixed

File: src/eko/compatibility.py
Comment (niclaurenti) [reply src/eko/compatibility.py:30]: Yes yes, but for example some tests have not it and kept failing. This is why I implemented it that way

Comment (AleCandido) [reply src/eko/compatibility.py:30]: "PTO" is always there, this is for free :)

File: src/eko/couplings.py
Comment (felixhekhorn) [inline src/eko/couplings.py:682]: You can copy the docstring

**Comment by niclaurenti** [comment]
@felixhekhorn @AleCandido Now everything should be fixed. Can we merge?

File: src/eko/compatibility.py
Comment (AleCandido) [reply src/eko/compatibility.py:29]: The best would be to have two decoupled methods (one for theory, the other for operators) that is not possible at the moment.

Most likely the best way would be to lift `ev_op_max_order` on theory side (and actually make it part of the `ModEv`). At that point you won't need any longer.

But ok, let's keep like this for the time being, and do it in a second step. What do you think @felixhekhorn?

File: src/eko/compatibility.py
Comment (AleCandido) [reply src/eko/compatibility.py:36]: Something like:

```py
def update_theory(theory):
    return update(theory, None)[0]
```

**Review by AleCandido — CHANGES_REQUESTED** [review]

File: src/eko/compatibility.py
Comment (felixhekhorn) [reply src/eko/compatibility.py:29]: > But ok, let's keep like this for the time being, and do it in a second step. What do you think @felixhekhorn?

definitely, this PR is already too big ...

**Review by AleCandido — APPROVED** [review]

**Comment by felixhekhorn** [comment]
fine to merge!

**Pull request merged**
